opencog/nlp/sureal/SuRealModule.cc

```diff
@@ -129,13 +124,106 @@ HandleSeq SuRealModule::do_sureal_match(Handle h)
         sVars.insert(n);
     }
 
-    logger().debug("[SuReal] starting pattern matcher");
+    // separate the disconnected clauses (this will happen often with SuReal)
+    std::set<HandleSeq> connectedClauses;
+    get_connected_components(sVars, qClauses, connectedClauses);
+
+    logger().debug("[SuReal] Found %d disconnected components", connectedClauses.size());
+
+    std::map<Handle, std::vector<std::map<Handle, Handle> > > collector;
+
+    // call the pattern matcher on each set of disconnected commponents
+    for (auto& c : connectedClauses)
```

Comment:
I did not study this carefully enough to see what you are trying to do, but .. wanted to point something out, that looks suspicious to me.

Suppose, for example there are 4 connected components, and the first one has M groundings, and the second has N, and the 3rd hand 4th have P, Q, then there are  M x N x P x Q total possibilities.  That requires 4 total nested for-loops to explore all the possibilities.  In general, if there are K components, you need K nested loops, i.e. you'd have to design something recursive to explore the possibilities ...  Seems to me that would have to apply here, as well, no?